code/WorkInProgress/DrMelonsStuff.dm

```diff
@@ -114,117 +147,166 @@
+				playsound(src.loc, "sound/misc/pourdrink.ogg", 60, 4)
+				src.on_reagent_change()
+				on = TRUE
+
+	proc/drain_bathtub(mob/user as mob)
+		src.add_fingerprint(user)
+		if (get_dist(usr, src) <= 1 && !is_incapacitated(usr))
+			if (src.reagents.total_volume)
+				user.visible_message("<span class='notice'>[user] reaches into the bath and pulls the plug.", "<span class='notice'>You reach into the bath and pull the plug.</span>")
+				if (ishuman(usr))
+					var/mob/living/carbon/human/H = usr
+					if(!H.gloves)
+						reagents.reaction(H, TOUCH)
+				playsound(src.loc, "sound/misc/drain_glug.ogg", 70, 1)
+				src.reagents.clear_reagents()
+				src.on_reagent_change()
+
+				var/count = 0
+				for (var/obj/O in src)
+					count++
+					qdel(O)
+				if (count > 0)
+					user.visible_message("<span class='alert'>...and something flushes down the drain. Damn!", "<span class='alert'>...and flush something down the drain. Damn!</span>")
+			else
+				boutput(usr, "<span class='notice'>The bathtub's already empty.</span>")
 
 	relaymove(mob/user as mob, dir)
-		src.eject_user()
-		boutput(user, "<span class='notice'>You get out of the bath.</span>")
-
+		src.eject_occupant(user)
 
 	process()
-		if (src.myuser)
-			if(src.myuser.loc != src)
-				src.myuser.pixel_y = 0
-				src.myuser = null
-				src.overlays = null
-				on_reagent_change()
-
+		if (src.on)
+			src.reagents.add_reagent(src.default_reagent, 100)
+			src.on_reagent_change()
+			if (src.reagents.is_full())
+				src.visible_message("<span class='notice'>As the [src] finishes filling, the tap shuts off automatically.</span>")
+				playsound(src.loc, "sound/misc/pourdrink2.ogg", 60, 5)
+				src.on = FALSE
+		if (src.occupant)
+			if(src.occupant.loc != src)
+				src.occupant.pixel_y = 0
+				src.occupant = null
+				src.on_reagent_change()
 				return
-			if(src.reagents.total_volume > 5)
-				reagents.reaction(src.myuser, TOUCH)
-				src.reagents.trans_to(src.myuser,5,1)
-			else
-				src.reagents.clear_reagents()
-				on_reagent_change()
-
-		return
+			if(src.reagents.total_volume)
+				src.reagents.reaction(src.occupant, TOUCH)
+				src.reagents.trans_to(src.occupant, 5)
 
 	MouseDrop_T(mob/living/carbon/human/target, mob/user)
-		if (src.myuser || !istype(target) || target.buckled || LinkBlocked(target.loc,src.loc) || BOUNDS_DIST(user, src) > 0 || BOUNDS_DIST(user, target) > 0 || is_incapacitated(user) || isAI(user))
-			return
-
-		var/msg
 
-		if(target == user && !user.stat)	// if drop self, then climbed in
-			msg = "[user.name] climbs into [src]."
-			boutput(user, "<span class='notice'>You climb into [src].</span>")
+		if (!istype(target) || target.buckled || LinkBlocked(target.loc,src.loc) || BOUNDS_DIST(user, src) > 1 || BOUNDS_DIST(user, target) > 1 || is_incapacitated(user) || isAI(user))
+			return
+		if (src.occupant)
+			boutput(user, "<span class='alert'>Someone's already in the [src]!</span>")
```

Comment:
i think you accidentally removed a return in here :)

Reply:
thank you!!